server/jwt/jwt_test.go

```diff
@@ -0,0 +1,212 @@
+//go:build all || unit
+// +build all unit
+
+package jwt
+
+//                                                                         __
+// .-----.-----.______.-----.----.-----.--.--.--.--.______.----.---.-.----|  |--.-----.
+// |  _  |  _  |______|  _  |   _|  _  |_   _|  |  |______|  __|  _  |  __|     |  -__|
+// |___  |_____|      |   __|__| |_____|__.__|___  |      |____|___._|____|__|__|_____|
+// |_____|            |__|                   |_____|
+//
+// Copyright (c) 2023 Fabio Cicerchia. https://fabiocicerchia.it. MIT License
+// Repo: https://github.com/fabiocicerchia/go-proxy-cache
+
+import (
+	"context"
+	"encoding/json"
+	"net/http/httptest"
+	"testing"
+
+	"github.com/fabiocicerchia/go-proxy-cache/config"
+	"github.com/lestrrat-go/jwx/jwt"
+	log "github.com/sirupsen/logrus"
+	"github.com/stretchr/testify/assert"
+)
+
+func TestAllowedScope(t *testing.T) {
+	co := &config.Jwt{Allowed_scopes: []string{"admin"}}
+	res := haveAllowedScope([]string{""}, co.Allowed_scopes)
+	assert.Equal(t, res, false, "No scope provided, should be false")
+
+	res = haveAllowedScope([]string{"admin"}, co.Allowed_scopes)
+	assert.Equal(t, res, true, "Admin is provided and allowed, should be true")
+
+	res = haveAllowedScope([]string{"admin"}, []string{})
+	assert.Equal(t, res, false, "No allowed scopes, should be false")
+
+	res = haveAllowedScope([]string{}, []string{})
+	assert.Equal(t, res, false, "Empty scopes and empty allowed scopes, should be false")
+}
+
+func TestGetScopesWithScopeClaim(t *testing.T) {
+	jwkKeySingle, _, _, _ := GenerateTestKeysAndKeySets()
+	strExpiredToken, _ := GenerateTestJWT(jwkKeySingle, "scope", true)
+
+	token, _ := jwt.ParseString(strExpiredToken, jwt.WithTypedClaim("scope", json.RawMessage{}))
+
+	res := getScopes(token)
+
+	assert.ElementsMatch(t, res, []string{"scope1", "scope2", "scope3"}, "Scopes provided doesn't match")
+}
+
+func TestGetScopesWithScpClaim(t *testing.T) {
+	jwkKeySingle, _, _, _ := GenerateTestKeysAndKeySets()
+	scpClaimToken, _ := GenerateTestJWT(jwkKeySingle, "scp", true)
+	token, _ := jwt.ParseString(scpClaimToken, jwt.WithTypedClaim("scp", json.RawMessage{}))
+
+	res := getScopes(token)
+
+	assert.ElementsMatch(t, res, []string{"scope1", "scope2", "scope3"}, "Scopes provided doesn't match")
+}
+
+func TestValidateJWT(t *testing.T) {
+	jwkKeySingle, jwkKeyMultiple, jsonJWKKeySetSingle, jsonJWKKeySetMultiple := GenerateTestKeysAndKeySets()
+	scpExpiredToken, _ := GenerateTestJWT(jwkKeyMultiple, "scp", true)
+	scopeGoodToken, _ := GenerateTestJWT(jwkKeySingle, "scope", false)
+	scopeGoodTokenMultiple, _ := GenerateTestJWT(jwkKeyMultiple, "scope", false)
+	scpGoodToken, _ := GenerateTestJWT(jwkKeySingle, "scp", false)
+	ts := CreateTestServer(t, jsonJWKKeySetSingle, jsonJWKKeySetMultiple)
+	defer ts.Close()
+
+	// Without any token
```

Comment:
would it be possible to split this test into multiple ones based on the scenario tested?